File: questions/218. The Skyline Problem/skyline.cpp
#include <vector>
#include <set>

using namespace std;

class Solution{
public:	
	vector<vector<int> > skyline(vector<vector<int> >& buildings){
		vector<pair<int, int> > events;
		vector<vector<int> > res;
		for(auto b: buildings){
			events.push_back(make_pair(b[0], -b[2]));
			events.push_back(make_pair(b[1], b[2]));
		}
		sort(events.begin(), events.end());
		multiset<int> mset;
		int cur, prev = 0;
		for(auto e: events){
			if(e.second < 0){
				mset.insert(-e.second);
			}else{
				mset.erase(mset.find(e.second));
			}
			cur = *mset.rbegin();
			if(cur != prev){
				prev = cur;
				res.push_back({e.first, cur});
			}
		}
		return res;
	}

};

File: questions/218. The Skyline Problem/skyline.py
class Solution(object):
	def skyline(self, buildings):
		events = [(L, -H, R) for L, R, H in buildings] + list({(R, 0, 0) for _, R, _ in buildings})
		events.sort()
		live = [(0,float("inf"))]
		res = [[0,0]]
		for Pos, NegH, R in events:
			while live[0][1] <= Pos: heappop(live)
			if NegH:
				heappush(live, (NegH, R))
			if res[-1][1] != -live[0][0]:
				res += [[Pos, -live[0][0]]]
		return res[1:]
